### PR TITLE
[rush-lib] Add project impact graph generator

### PR DESCRIPTION
**Pull request by wesley-weiming**
## Summary

This PR is related to [Project-Impact-Graph](https://github.com/tiktok/project-impact-graph/tree/main)
It adds a project-impact-graph.yaml file generator to rush-lib

## How it was tested

1. run `rushx build` under `rushstack/libraries/rush-lib`
2. Configure `"useProjectImpactGraphGeneratorAfterRushUpdate": true` in the `experiments.json` file
3. After executing `rush update`, rush will automatically generate `project-impact-graph.yaml` file in repository root path

### CODE REVIEW

**Comment by wesley-weiming** [comment]
> @wesley-weiming please read the following Contributor License Agreement(CLA). If you agree with the CLA, please reply with the following information.
> 
> ```
> @microsoft-github-policy-service agree [company="{your company}"]
> ```
> 
> > Options:
> > 
> > * (default - no company specified) I have sole ownership of intellectual property rights to my Submissions and I am not making Submissions in the course of work for my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 
> > * (when company given) I am making Submissions in the course of work for my employer (or my employer has intellectual property rights in my Submissions by contract or applicable law). I have permission from my employer to make Submissions and enter into this Agreement on behalf of my employer. By signing below, the defined term “You” includes me and my employer.
> > 
> > ```
> > @microsoft-github-policy-service agree company="Microsoft"
> > ```
> 
> Contributor License Agreement

@microsoft-github-policy-service agree company="TikTok"

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import fs from 'fs';
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:4]: Use `FileSystem` from `node-core-library`.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+  projects: {
+    [key: string]: IProjectConfiguration;
+  };
+}
+
+export class ProjectImpactGraphGenerator {
+  /**
+   * Full path of repository root
+   */
+  private _repositoryRoot: string;
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:35]:
```suggestion
  private readonly _repositoryRoot: string;
```

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+
+  /**
+   * Projects within the Rush configuration
+   */
+  private _projects: RushConfigurationProject[];
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:40]:
```suggestion
  private readonly _projects: RushConfigurationProject[];
```

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+    this._projects = rushConfiguration.projects;
+  }
+
+  /**
+   * Load global excluded globs
+   * @param repositoryRoot
+   */
+  private _loadGlobalExcludedGlobs(repositoryRoot: string): string[] | undefined {
+    const filePath: string = path.join(repositoryRoot, '.mergequeueignore');
+    if (fs.existsSync(filePath)) {
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:62]: Can you do filesystem ops async?

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+        const projectExcludedGlobs: string[] | undefined = this._loadProjectExcludedGlobs(
+          project.projectRelativeFolder
+        );
+        if (projectExcludedGlobs) {
+          content.projects[`${project.packageName}`].excludedGlobs = projectExcludedGlobs;
+        }
+      }
+    });
+
+    fs.writeFileSync(path.join(this._repositoryRoot, 'project-impact-graph.yaml'), yaml.safeDump(content));
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:109]: Seems like this file path should be configurable.

Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:109]:
```suggestion
    fs.writeFileSync(`${this._repositoryRoot}/project-impact-graph.yaml`, yaml.safeDump(content));
```

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      return globs.map((glob) => path.join(projectRootRelativePath, glob));
+    }
+  }
+
+  /**
+   * Core Logic: generate project-impact-graph.yaml
+   */
+  public generate(): void {
+    const stopwatch: Stopwatch = Stopwatch.start();
+    const content: IFileSchema = {} as IFileSchema;
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:85]: If you generate the individual properties first, you can construct this without the cast.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      return globs;
+    }
+  }
+
+  /**
+   * Load project excluded globs
+   * @param projectRootRelativePath - project root relative path
+   */
+  private _loadProjectExcludedGlobs(projectRootRelativePath: string): string[] | undefined {
+    const filePath: string = path.join(this._repositoryRoot, projectRootRelativePath, '.mergequeueignore');
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:73]:
```suggestion
    const filePath: string = `${this._repositoryRoot}/${projectRootRelativePath}/.mergequeueignore`;
```

Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:73]: Stick `.mergequeueignore` in a const in RushConstants.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 8]
```diff
@@ -0,0 +1,114 @@
...
+    if (fs.existsSync(filePath)) {
+      const globs: string[] = fs.readFileSync(filePath).toString().split('\n');
+      return globs.map((glob) => path.join(projectRootRelativePath, glob));
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:76]: This file is normal globs, and not in `.gitignore` format?

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 9]
```diff
@@ -0,0 +1,114 @@
...
+    this._repositoryRoot = rushConfiguration.rushJsonFolder;
+    this._projects = rushConfiguration.projects;
+  }
+
+  /**
+   * Load global excluded globs
+   * @param repositoryRoot
+   */
+  private _loadGlobalExcludedGlobs(repositoryRoot: string): string[] | undefined {
+    const filePath: string = path.join(repositoryRoot, '.mergequeueignore');
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:61]:
```suggestion
    const filePath: string = `${repositoryRoot}/.mergequeueignore`;
```

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+  /**
+   * Projects within the Rush configuration
+   */
+  private _projects: RushConfigurationProject[];
+
+  /**
+   * Default global excluded globs
+   * Only used if the `<repository_root>/.mergequeueignore` does not exist
+   */
+  private _DefaultGlobalExcludedGlobs: string[] = ['common/autoinstallers/**'];
```
Comment (iclanton) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:46]: Stick this in a const outside the class.

File: libraries/rush-lib/src/schemas/experiments.schema.json [hunk 11]
```diff
@@ -53,6 +53,10 @@
     "forbidPhantomResolvableNodeModulesFolders": {
       "description": "If true, Rush will not allow node_modules in the repo folder or in parent folders.",
       "type": "boolean"
+    },
+    "useProjectImpactGraphGeneratorAfterRushUpdate": {
```
Comment (iclanton) [inline libraries/rush-lib/src/schemas/experiments.schema.json:57]: Update the `rush init` template.

**Review by iclanton — CHANGES_REQUESTED** [review]

**Comment by iclanton** [comment]
Are users expected to check this file in?

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+   */
+  public generate(): void {
+    const stopwatch: Stopwatch = Stopwatch.start();
+    const content: IFileSchema = {} as IFileSchema;
+    content.globalExcludedGlobs =
+      this._loadGlobalExcludedGlobs(this._repositoryRoot) || this._DefaultGlobalExcludedGlobs;
+    content.projects = {};
+    this._projects.forEach((project) => {
+      // ignore the top project
+      if (project.projectRelativeFolder !== '.') {
```
Comment (EscapeB) [inline libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:91]: Why top project is ignored in this forEach loop?

Top level package.json/project is not recommended in Rush.js, but if we do have such a package.json/project at the top level, any code changes in top level project may affect all projects inside Monorepo and we actually need to keep it.

In my opinion, if we need to ignore some top level changes, we can use `.mergeignore` like other normal projects. And for `GlobalExcludedGlobs`, maybe we can find another place to store it.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 5 above)
Comment (octogonz) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts]: @iclanton The YAML file is meant to be read directly from Git by external systems such as GitHub or Mergify, which are language agnostic. This is similar to other files such as `CODEOWNERS` or `.gitattributes`.  We can make it configurable, but then those integrating systems would need some way to discover or store the file path. And is it merely the location that is configurable, but the filename part is standard? What do you think?

If the location is in the repository root, then maybe the filename should be `.project-impact-graph.yaml`.

File: libraries/rush-lib/src/api/ExperimentsConfiguration.ts [hunk 13]
```diff
@@ -76,6 +76,11 @@ export interface IExperimentsJson {
    * If true, Rush will not allow node_modules in the repo folder or in parent folders.
    */
   forbidPhantomResolvableNodeModulesFolders?: boolean;
+
+  /**
+   * If true, Rush will generate project-impact-graph.yaml at repositoryRoot after `rush update` finished.
```
[the remark this answers is not in the document]
Comment (octogonz) [reply libraries/rush-lib/src/api/ExperimentsConfiguration.ts:81]: The specification is meant to be language agnostic, so we cannot assume that the repo will be using Rush or even PNPM.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 12 above)
Comment (octogonz) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:91]: `for (const project of this._projects) {` is more idiomatic and avoids a closure.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 9 above)
Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:61]: I'm not convinced anyone actually cares if paths have mixed slashes.

`path.join` is absolutely more expensive than a simple string concatenation, and it doesn't offer any non-aesthetic benefits if we know the path doesn't contain any `..` or `.` folders.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 8 above)
Comment (wesley-weiming) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:76]: > This file is normal globs, and not in `.gitignore` format?

It's normal globs, and has the same format as `.gitignore`

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 12 above)
Comment (wesley-weiming) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:91]: > Why top project is ignored in this forEach loop?
> 
> Top level package.json/project is not recommended in Rush.js, but if we do have such a package.json/project at the top level, any code changes in top level project may affect all projects inside Monorepo and we actually need to keep it.
> 
> In my opinion, if we need to ignore some top level changes, we can use `.mergeignore` like other normal projects. And for `GlobalExcludedGlobs`, maybe we can find another place to store it.

According to the specification  of includedGlobs, top project will cover the whole repo's files, all changed files 
 will be identified as having a dependency. such as:
```
projects:
  top-project:
    includedGlobs:
      - ./**
```

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,121 @@
...
+          projects[`${project.packageName}`].excludedGlobs = projectExcludedGlobs;
+        }
+      }
+    }
+    const content: IFileSchema = { globalExcludedGlobs, projects };
+
+    FileSystem.writeFile(
+      path.join(this._repositoryRoot, 'project-impact-graph.yaml'),
+      yaml.safeDump(content)
+    );
```
[the remark this answers is not in the document]
Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:116]: Also stick the YAML filename in a constant in RushConstsnts.

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+    const stopwatch: Stopwatch = Stopwatch.start();
+
+    const globalExcludedGlobs: string[] =
+      this._loadGlobalExcludedGlobs(this._repositoryRoot) || DefaultGlobalExcludedGlobs;
+    const projects: {
+      [key: string]: IProjectConfiguration;
+    } = {};
+    for (const project of this._projects) {
+      // ignore the top project
+      if (project.projectRelativeFolder !== '.') {
```
[the remark this answers is not in the document]
Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts:94]: (By "project," I mean a project specified in rush.json, not a package.json file. A root-level package.json is technically allowed, but it doesn't do anything in a Rush repo)

File: libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts (hunk 5 above)
Comment (iclanton) [reply libraries/rush-lib/src/logic/ProjectImpactGraphGenerator.ts]: I like that name as well.

**Review by iclanton — APPROVED** [review]

**Pull request merged**